### PR TITLE
`azurerm_network_interface` - Support `auxiliary_mode`, `auxiliary_sku`

### PR DESCRIPTION
**Pull request by ms-zhenhua**
## Swagger Link

https://github.com/Azure/azure-rest-api-specs/blob/1df6d6f671dc5059016fbc2c0a624e01f0b2972c/specification/network/resource-manager/Microsoft.Network/stable/2023-02-01/networkInterface.json#L1212

https://github.com/Azure/azure-rest-api-specs/blob/1df6d6f671dc5059016fbc2c0a624e01f0b2972c/specification/network/resource-manager/Microsoft.Network/stable/2023-02-01/networkInterface.json#L1226C10-L1226C22

## Test Result

image

=== RUN   TestAccNetworkInterface_auxiliary
=== PAUSE TestAccNetworkInterface_auxiliary
=== CONT  TestAccNetworkInterface_auxiliary
--- PASS: TestAccNetworkInterface_auxiliary (360.05s)
PASS

### CODE REVIEW

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hi @ms-zhenhua 

I've taken a look through and left some comments inline, however many of the comments I've raised in this PR have been raised in previous PR's (for example the `None` issue, [which can be found here](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/schema-design-considerations.md#the-none-value-or-similar)) - as such would you mind fixing these up and paying particular attention to the `None` issue going forward?

Thanks!

File: internal/services/network/network_interface_resource.go [hunk 1]
```diff
@@ -126,6 +126,32 @@ func resourceNetworkInterface() *pluginsdk.Resource {
...
 			// Optional
+			"auxiliary_mode": {
+				Type:     pluginsdk.TypeString,
+				Optional: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(networkinterfaces.NetworkInterfaceAuxiliaryModeAcceleratedConnections),
+					string(networkinterfaces.NetworkInterfaceAuxiliaryModeFloating),
+					string(networkinterfaces.NetworkInterfaceAuxiliaryModeNone),
+				}, false),
+				Default:      string(networkinterfaces.NetworkInterfaceAuxiliaryModeNone),
```
[the remark this answers is not in the document]
Comment (ms-zhenhua) [reply internal/services/network/network_interface_resource.go]: code updated

File: internal/services/network/network_interface_resource.go [hunk 2]
```diff
@@ -126,6 +126,32 @@ func resourceNetworkInterface() *pluginsdk.Resource {
...
+				Type:     pluginsdk.TypeString,
+				Optional: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(networkinterfaces.NetworkInterfaceAuxiliarySkuAEight),
+					string(networkinterfaces.NetworkInterfaceAuxiliarySkuAFour),
+					string(networkinterfaces.NetworkInterfaceAuxiliarySkuAOne),
+					string(networkinterfaces.NetworkInterfaceAuxiliarySkuATwo),
+					string(networkinterfaces.NetworkInterfaceAuxiliarySkuNone),
+				}, false),
+				Default:      string(networkinterfaces.NetworkInterfaceAuxiliarySkuNone),
```
[the remark this answers is not in the document]
Comment (ms-zhenhua) [reply internal/services/network/network_interface_resource.go]: code updated

File: internal/services/network/network_interface_resource.go [hunk 3]
```diff
@@ -462,6 +512,8 @@ func resourceNetworkInterfaceRead(d *pluginsdk.ResourceData, meta interface{}) e
 				return fmt.Errorf("setting `applied_dns_servers`: %+v", err)
 			}
 
+			d.Set("auxiliary_mode", pointer.From(props.AuxiliaryMode))
+			d.Set("auxiliary_sku", pointer.From(props.AuxiliarySku))
```
[the remark this answers is not in the document]
Comment (ms-zhenhua) [reply internal/services/network/network_interface_resource.go]: code updated

File: internal/services/network/network_interface_resource_test.go [hunk 4]
```diff
@@ -388,6 +416,43 @@ resource "azurerm_network_interface" "test" {
...
+	// To not affect other testcases of `Network`, hard-code to that for now
+	data.Locations.Primary = "westus"
+
+	if mode != "" {
+		mode = fmt.Sprintf(`auxiliary_mode = "%s"`, mode)
+	}
+
+	if sku != "" {
+		sku = fmt.Sprintf(`auxiliary_sku = "%s"`, sku)
+	}
```
[the remark this answers is not in the document]
Comment (ms-zhenhua) [reply internal/services/network/network_interface_resource_test.go]: code updated

File: website/docs/r/network_interface.html.markdown [hunk 5]
```diff
@@ -60,6 +60,12 @@ The following arguments are supported:
 
 ---
 
+* `auxiliary_mode` - (Optional) The Auxiliary mode of the Network Interface. Possible values are `AcceleratedConnections`, `Floating` and `None`. Defaults to `None`.
+
+-> **Note:** This field requires the preview feature is enabled. See the [Prerequisites](https://learn.microsoft.com/en-us/azure/networking/nva-accelerated-connections#prerequisites) for more details.
+
+* `auxiliary_sku` - (Optional) The Auxiliary SKU of the Network Interface. Possible values are `A1`, `A2`, `A4`, `A8` and `None`. Defaults to `None`.
```
[the remark this answers is not in the document]
Comment (ms-zhenhua) [reply website/docs/r/network_interface.html.markdown]: firstly: Yes, I updated the doc for both of them
secondly: doc updated.
thirdly: It is used for enabling Accelerated Connections to improve networking performance. I have updated the doc.

File: website/docs/r/network_interface.html.markdown [hunk 6]
```diff
@@ -60,6 +60,12 @@ The following arguments are supported:
 
 ---
 
+* `auxiliary_mode` - (Optional) The Auxiliary mode of the Network Interface. Possible values are `AcceleratedConnections`, `Floating` and `None`. Defaults to `None`.
```
[the remark this answers is not in the document]
Comment (ms-zhenhua) [reply website/docs/r/network_interface.html.markdown]: firstly: doc updated.
secondly: it is used for tuning the performance of network connections. I have updated the doc

**Comment by ms-zhenhua** [comment]
Hi @tombuildsstuff,

Thank you for your review. Sorry for not noticing the latest shema design. I will pay more attention in future. I have updated all comments and rerun the tests. Kindly take another review.

Thanks.

File: internal/services/network/network_interface_resource.go [hunk 7]
```diff
@@ -126,6 +126,28 @@ func resourceNetworkInterface() *pluginsdk.Resource {
...
+
+			"auxiliary_sku": {
+				Type:     pluginsdk.TypeString,
+				Optional: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(networkinterfaces.NetworkInterfaceAuxiliarySkuAEight),
+					string(networkinterfaces.NetworkInterfaceAuxiliarySkuAFour),
+					string(networkinterfaces.NetworkInterfaceAuxiliarySkuAOne),
+					string(networkinterfaces.NetworkInterfaceAuxiliarySkuATwo),
+				}, false),
```
Comment (tombuildsstuff) [inline internal/services/network/network_interface_resource.go:147]: we can use:

```suggestion
				ValidateFunc: validation.StringInSlice(networkinterfaces.PossibleValuesForNetworkInterfaceAuxiliarySku(), false),
```

File: internal/services/network/network_interface_resource.go [hunk 8]
```diff
@@ -126,6 +126,28 @@ func resourceNetworkInterface() *pluginsdk.Resource {
 			},
 
 			// Optional
+			"auxiliary_mode": {
+				Type:     pluginsdk.TypeString,
+				Optional: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(networkinterfaces.NetworkInterfaceAuxiliaryModeAcceleratedConnections),
+					string(networkinterfaces.NetworkInterfaceAuxiliaryModeFloating),
```
Comment (tombuildsstuff) [inline internal/services/network/network_interface_resource.go:134]: we can use `possiblevaluesfor` here fwiw

File: internal/services/network/network_interface_resource.go [hunk 9]
```diff
@@ -325,6 +355,22 @@ func resourceNetworkInterfaceUpdate(d *pluginsdk.ResourceData, meta interface{})
 		},
 	}
 
+	if d.HasChange("auxiliary_mode") {
+		if auxiliaryMode, hasAuxiliaryMode := d.GetOk("auxiliary_mode"); hasAuxiliaryMode {
+			update.Properties.AuxiliaryMode = pointer.To(networkinterfaces.NetworkInterfaceAuxiliaryMode(auxiliaryMode.(string)))
+		}
+	} else {
+		update.Properties.AuxiliaryMode = existing.Model.Properties.AuxiliaryMode
+	}
```
Comment (tombuildsstuff) [inline internal/services/network/network_interface_resource.go:364]: how do users unset this property?

File: internal/services/network/network_interface_resource.go [hunk 10]
```diff
@@ -325,6 +355,22 @@ func resourceNetworkInterfaceUpdate(d *pluginsdk.ResourceData, meta interface{})
...
+
+	if d.HasChange("auxiliary_sku") {
+		if auxiliarySku, hasAuxiliarySku := d.GetOk("auxiliary_sku"); hasAuxiliarySku {
+			update.Properties.AuxiliarySku = pointer.To(networkinterfaces.NetworkInterfaceAuxiliarySku(auxiliarySku.(string)))
+		}
+	} else {
+		update.Properties.AuxiliarySku = existing.Model.Properties.AuxiliarySku
+	}
```
Comment (tombuildsstuff) [inline internal/services/network/network_interface_resource.go:372]: how do users unset this property?

File: internal/services/network/network_interface_resource_test.go [hunk 11]
```diff
@@ -388,6 +416,62 @@ resource "azurerm_network_interface" "test" {
...
+  enable_accelerated_networking = true
+
+  ip_configuration {
+    name                          = "primary"
+    subnet_id                     = azurerm_subnet.test.id
+    private_ip_address_allocation = "Dynamic"
+  }
+
+  tags = {
+    fastpathenabled = "true"
```
Comment (tombuildsstuff) [inline internal/services/network/network_interface_resource_test.go:440]: what's this?

File: internal/services/network/network_interface_resource_test.go [hunk 12]
```diff
@@ -388,6 +416,62 @@ resource "azurerm_network_interface" "test" {
...
```
Comment (tombuildsstuff) [inline internal/services/network/network_interface_resource_test.go:469]: what's this?

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hi @ms-zhenhua 

Thanks for pushing those changes - I've taken a look through and left some comments inline, however would it be possible to explain the `fastpathenabled` tag?

Thanks!

File: internal/services/network/network_interface_resource.go (hunk 7 above)
Comment (ms-zhenhua) [reply internal/services/network/network_interface_resource.go]: code updated. Thanks.

File: internal/services/network/network_interface_resource.go (hunk 8 above)
Comment (ms-zhenhua) [reply internal/services/network/network_interface_resource.go]: code updated. Thanks.

File: internal/services/network/network_interface_resource.go (hunk 9 above)
Comment (ms-zhenhua) [reply internal/services/network/network_interface_resource.go:356]: If users remove this property from the .tf, `hasAuxiliaryMode = false` so that `update.Properties.AuxiliaryMode` will not be assigned a value, then Azure will unset this property.

File: internal/services/network/network_interface_resource.go (hunk 10 above)
Comment (ms-zhenhua) [reply internal/services/network/network_interface_resource.go:364]: If users remove this property from the .tf, `hasAuxiliarySku = false` so that `update.Properties.AuxiliarySku` will not be assigned a value, then Azure will unset this property.

File: internal/services/network/network_interface_resource_test.go (hunk 11 above)
Comment (ms-zhenhua) [reply internal/services/network/network_interface_resource_test.go:440]: It's a custom tag which is mentioned in https://learn.microsoft.com/en-us/azure/networking/nva-accelerated-connections#prerequisites and is used to enable `AuxiliaryMode`. Users will get detailed guidance after signing up for this feature mentioned in https://learn.microsoft.com/en-us/azure/networking/nva-accelerated-connections#prerequisites

File: internal/services/network/network_interface_resource_test.go (hunk 12 above)
Comment (ms-zhenhua) [reply internal/services/network/network_interface_resource_test.go:469]: Answered above.

**Comment by ms-zhenhua** [comment]
Hi @tombuildsstuff, thank you for your review. I have resolved the new comments and explained `fastpathenabled` tag. Could you kindly take another review? Thanks.

## Test Result
=== RUN   TestAccNetworkInterface_auxiliary
=== PAUSE TestAccNetworkInterface_auxiliary
=== CONT  TestAccNetworkInterface_auxiliary
--- PASS: TestAccNetworkInterface_auxiliary (524.86s)
PASS

File: internal/services/network/network_interface_resource.go (hunk 10 above)
Comment (tombuildsstuff) [reply internal/services/network/network_interface_resource.go:364]: However the Go SDK for this defines this field as `omitempty` (due to it being optional in the API)

> AuxiliarySku                *NetworkInterfaceAuxiliarySku       `json:"auxiliarySku,omitempty"`

Meaning that if it's `nil` no value will be sent to the API - meaning this can't be cleared out. So can we add a test for this? This may need to be explicitly set to an empty string to handle this use-case

Comment (ms-zhenhua) [reply internal/services/network/network_interface_resource.go:364]: The testcase has already existed in this PR and it can pass. In the testcase, the workflow is [unset]-> [set] -> [unset].

image

File: website/docs/r/network_interface.html.markdown [hunk 13]
```diff
@@ -60,6 +60,12 @@ The following arguments are supported:
 
 ---
 
+* `auxiliary_mode` - (Optional) Specifies auxiliary mode for enabling [Accelerated Connections](https://learn.microsoft.com/en-us/azure/networking/nva-accelerated-connections) to improve networking performance. Possible values are `AcceleratedConnections` and `Floating`.
+
+* `auxiliary_sku` - (Optional) Specifies auxiliary SKU for tuning the performance of network connections. Possible values are `A1`, `A2`, `A4` and `A8`.
```
Comment (tombuildsstuff) [inline website/docs/r/network_interface.html.markdown:65]: (as above)

File: website/docs/r/network_interface.html.markdown [hunk 14]
```diff
@@ -60,6 +60,12 @@ The following arguments are supported:
...
+
+-> **Note:** `auxiliary_mode` and `auxiliary_sku` require the preview feature is enabled. See the [Prerequisites](https://learn.microsoft.com/en-us/azure/networking/nva-accelerated-connections#prerequisites) for more details.
```
Comment (tombuildsstuff) [inline website/docs/r/network_interface.html.markdown:67]: The Preview note should be listed under every preview field, and something along the lines of:

```suggestion
-> **Note:** `auxiliary_mode` is in **Preview** and requires that the preview is enabled - [more information can be found in the Azure documentation](https://learn.microsoft.com/azure/networking/nva-accelerated-connections#prerequisites).
```

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
Left a few comments around the documentation, as we should have a succinct explanation of what this does without requiring clicking through to another site, but if we can fix this up then it should otherwise be good to go 👍

File: website/docs/r/network_interface.html.markdown [hunk 15]
```diff
@@ -60,6 +60,12 @@ The following arguments are supported:
 
 ---
 
+* `auxiliary_mode` - (Optional) Specifies auxiliary mode for enabling [Accelerated Connections](https://learn.microsoft.com/en-us/azure/networking/nva-accelerated-connections) to improve networking performance. Possible values are `AcceleratedConnections` and `Floating`.
```
[the remark this answers is not in the document]
Comment (ms-zhenhua) [reply website/docs/r/network_interface.html.markdown]: Doc updated. Thanks.

File: website/docs/r/network_interface.html.markdown (hunk 13 above)
Comment (ms-zhenhua) [reply website/docs/r/network_interface.html.markdown]: Doc updated. Thanks.

File: website/docs/r/network_interface.html.markdown (hunk 14 above)
Comment (ms-zhenhua) [reply website/docs/r/network_interface.html.markdown]: Doc updated. Thanks.

**Comment by ms-zhenhua** [comment]
Hi @tombuildsstuff, thank you for another review. I have updated the doc. Kindly help have a check. Thanks.
